using custom request types for sitemap spider

codinguncut opened the pull request:
I am deriving from SitemapSpider for exploratory crawls and only want to see sitemap links, without necessarily following them.
For this reason it's important to be able to differentiate whether a Request is from a SitemapIndex or from a UrlSet.
I have derived (named) classes from Request to make this easily discoverable in derived classes and/or pipeline.

kmike commented:
The feature sounds reasonable!

Previously the way to do it would be to add something to request.meta. But since https://github.com/scrapy/scrapy/pull/2510 (included in scrapy 1.3.2) custom request subclasses should also work. This PR can be a good chance to dogfood these custom request classes. So the implementation also looks fine to me.

Could you please add docs and tests for this feature? Existing docs: https://github.com/scrapy/scrapy/blob/master/docs/topics/spiders.rst and tests: https://github.com/scrapy/scrapy/blob/afac3fd2c28ac5b16e7e3c95fca5acd1334f8c58/tests/test_spider.py#L293

kmike commented on scrapy/spiders/sitemap.py:
```diff
@@ -9,6 +9,10 @@
+# differentiating requests for use in child classes
+class SitemapIndexRequest(Request): pass
+class UrlSetRequest(Request): pass
+class RobotsRequest(Request): pass
```
Could you please use boring pep8 style (`pass` on a new line, two lines between classes)? 

If you leave it as-is then over the next year or two someone will inevitably submit a pull request to fix code style; that'd be just a time waste for nothing :)

codinguncut replied:
sorry, lost track of this PR. will do docs, test and pep8 compliance

kmike commented:
yet another way to implement this: https://github.com/scrapy/scrapy/pull/2082

codinguncut commented:
Hi,

I have rebased my branch, made code changes PEP8 compliant and added minimal documentation.

I think I messed up the branch/pull request because of the rebase from upstream scrapy. Maybe it would make sense to reopen this with a fresh branch/commit...

I was a bit unsure how to test this, and also since all I'm doing is to replace a class with it's child class I'm not sure how much testing is necessary at least for backward compatibility/regression.

I'd be grateful if you could merge this...

Johannes
